target/cheri-common/cheri-helper-utils.h

```diff
@@ -249,14 +249,66 @@ static inline uint32_t perms_for_store(CPUArchState *env, uint32_t cs)
 typedef void QEMU_NORETURN (*unaligned_memaccess_handler)(CPUArchState *env,
                                                           target_ulong addr,
                                                           uintptr_t retpc);
-// Do all the permission and bounds checks for loads/stores on cbp.
-// Use perms_for_load() and perms_for_store() for required_perms.
-target_ulong cap_check_common_reg(uint32_t required_perms, CPUArchState *env,
-                                  uint32_t cb, target_ulong offset,
-                                  uint32_t size, uintptr_t _host_return_address,
-                                  const cap_register_t *cbp,
-                                  uint32_t alignment_required,
-                                  unaligned_memaccess_handler unaligned_handler);
+/* Do all the permission and bounds checks for loads/stores on cbp.
+ * Use perms_for_load() and perms_for_store() for required_perms.
+ * Note: This is marked as QEMU_ALWAYS_INLINE since profiling indicates it is
+ * called rather frequently. Not removing dead branches/propagating
+ * the constant alignment argument has a noticeable performance impact:
+ * we were generating an x86 div instruction to check non-constant alignment
```

Comment:
Presumably that's no longer true now we use P2 and that, whilst that would've likely also constant-propagated its way through to the `%` when inlined, now we use the builtin/bitmasking it should still be less bad?

Reply:
Yes the first commit removes the `div` instruction but does not constant propagate (314s -> 288s), the second does the constant propagation (288s -> 277s).